### DISCUSSION TITLE
Discourage using `is_...` stuff for trivial type checks

### DISCUSSION THREAD

**Discussion by sylee957** [sympy/sympy#24554] [General]
There are many properties listed here

```python3
    is_number = False
    is_Atom = False
    is_Symbol = False
    is_symbol = False
    is_Indexed = False
    is_Dummy = False
    is_Wild = False
    is_Function = False
    is_Add = False
    is_Mul = False
    is_Pow = False
    is_Number = False
    is_Float = False
    is_Rational = False
    is_Integer = False
    is_NumberSymbol = False
    is_Order = False
    is_Derivative = False
    is_Piecewise = False
    is_Poly = False
    is_AlgebraicNumber = False
    is_Relational = False
    is_Equality = False
    is_Boolean = False
    is_Not = False
    is_Matrix = False
    is_Vector = False
    is_Point = False
    is_MatAdd = False
    is_MatMul = False
```

Unfortunately, I don't see the point of `is_Symbol`, `is_MatAdd` to be better than `isinstance`,
especially after mypy or some IDE are able to infer the types from `isinstance`, while it can not from `is_...` stuff.

So I'd think of at least those should be slowly removed by someone who wants to make changes to the code styles.

Some of the checks, like `is_symbol` are fine or hard to remove because they are designed to bring semantical type check.
But anyway, the duplicate of `is_Symbol` and `is_symbol` just confuses users, and we are not even sure if something capitalized follows strict rule altogether,

**Comment by oscarbenjamin** [comment]
Broadly I would like to get rid of these but two points to be careful of here are that:

1. The `is_` checks often break what would otherwise be cyclic imports which are very problematic in SymPy e.g. `isinstance(obj, Add)` requires importing `Add` whereas `is_Add` does not.
2. Checkng a static `is_` attribute is faster so there could be a performance hit in some cases.

Timings for `is_`:
```python
In [31]: %timeit isinstance(x, Number)
203 ns ± 1.92 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [32]: %timeit x.is_Number
42.4 ns ± 0.443 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
When thinking about these timings particular attention should be paid to those used in the critical loops like `Add/Mul.flatten` or `Pow.__new__`.

I think that `is_Symbol` and `is_symbol` are not completely equivalent i.e. some classes have different values for the two. I seem to remember having a problem with this at some point.

Note that `is_number` is completely different from the rest. I agree that it is badly named though. A clearer name would be something like `can_evalf_complex`. Likewise `is_comparable` could be `can_evalf_real`.

Certainly many of these are not widely used and we could just stop using them. Unfortunately the most common ones are also the ones that are most likely to be significant for both circular import problems and also performance. Some counts:
```console
$ git grep is_Atom | wc -l
81
$ git grep is_Symbol | wc -l
125
$ git grep is_symbol | wc -l
95
$ git grep is_Indexed | wc -l
2
$ git grep is_Dummy | wc -l
7
$ git grep is_Wild | wc -l
4
$ git grep is_Function | wc -l
60
$ git grep is_Add | wc -l
310
```
I guess some like `is_Indexed` will be very easy to remove but others like `is_Add` would be more problematic.

> So I'd think of at least those should be slowly removed by someone who wants to make changes to the code styles.

My understanding of this is that you mean that we should change the codebase to stop using these attributes. Is that correct?

Removing the attributes would be a significant backwards compatibility problem and I don't think I would support that. Retrospectively it would be better if these attributes weren't there because it's clearly an unscalable approach but I don't think that we should remove or even deprecated (most of) them.

**Reply by sylee957** [reply]
No I'd refer that by keeping the stuff only in `Basic` but removing the usage cases throughout sympy codebase

**Comment by asmeurer** [comment]
Yeah, I don't like the `is_` attributes that are just there for type checks. This was added a long time ago as a micro optimization because it's faster than isinstance. But I honestly don't think it's worth the confusion, especially since the exact same naming is also used for the assumptions and other attributes. And it goes against standard Python practices. I agree we should not use them and probably recommend against them, if not completely deprecate and remove them.

**Comment by sylee957** [comment]
I wonder that `isinstance` can be made faster in more uniform way, like cacheing, but probably there is no way to get out of this.
